src/WebJobs.Script.WebHost/FileMonitoringService.cs

```diff
@@ -126,12 +117,50 @@ private void InitializeFileWatchers()
                     .Subscribe((msg) => ScheduleRestartAsync(false)
                     .ContinueWith(t => _logger.LogCritical(t.Exception.Message),
                         TaskContinuationOptions.OnlyOnFaulted | TaskContinuationOptions.ExecuteSynchronously)));
+
+            // Delay starting up for logging and debug file watchers to avoid long start up times
+            Utility.ExecuteAfterDelay(InitializeNonBlockingFileWatchers, TimeSpan.FromSeconds(_delayedWatchersDelaySeconds));
+        }
+
+        /// <summary>
+        /// Initializes the file and directory monitoring that does not need to happen as part of a Host startup
+        /// These watchers can be started after a delay to avoid startup performance issue
+        /// </summary>
+        private void InitializeNonBlockingFileWatchers()
+        {
+            if (_watchersStopped)
+            {
+                return;
+            }
+
+            lock (_stopWatchersLock)
+            {
+                // We need this duplication to avoid race if watchers were stopped just when we acquired the lock
+                if (!_watchersStopped)
+                {
+                    _debugModeFileWatcher = new AutoRecoveringFileSystemWatcher(_hostLogPath, ScriptConstants.DebugSentinelFileName,
+                            includeSubdirectories: false, changeTypes: WatcherChangeTypes.Created | WatcherChangeTypes.Changed);
+                    _debugModeFileWatcher.Changed += OnDebugModeFileChanged;
+                    _logger.LogDebug("Debug file watch initialized.");
+
+                    _diagnosticModeFileWatcher = new AutoRecoveringFileSystemWatcher(_hostLogPath, ScriptConstants.DiagnosticSentinelFileName,
+                           includeSubdirectories: false, changeTypes: WatcherChangeTypes.Created | WatcherChangeTypes.Changed);
+                    _diagnosticModeFileWatcher.Changed += OnDiagnosticModeFileChanged;
+                    _logger.LogDebug("Diagnostic file watch initialized.");
+                }
+            }
         }
 
         private void StopFileWatchers()
         {
+            if (_watchersStopped)
+            {
+                return;
+            }
+
             lock (_stopWatchersLock)
             {
+                // We need this duplication to avoid race if watchers were stopped just when we acquired the lock
```

Comment:
nit: Not sure this comment is needed. This is a pretty common pattern.

Reply:
oh interesting. I wasn't aware of this pattern with duplicate checking. Ok, I will remove it. But I am going to make a separate PR just to avoid blocking this on that (waiting for builds, another review etc).

Reply:
https://github.com/Azure/azure-functions-host/pull/5645